src/main/java/com/google/devtools/build/lib/util/Fingerprint.java

```diff
@@ -14,43 +14,36 @@
 
 package com.google.devtools.build.lib.util;
 
-import com.google.common.io.ByteStreams;
+import com.google.common.hash.Funnels;
+import com.google.common.hash.HashFunction;
+import com.google.common.hash.Hasher;
 import com.google.devtools.build.lib.vfs.DigestHashFunction;
 import com.google.devtools.build.lib.vfs.Path;
 import com.google.devtools.build.lib.vfs.PathFragment;
 import com.google.protobuf.ByteString;
 import com.google.protobuf.CodedOutputStream;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
-import java.security.DigestException;
-import java.security.DigestOutputStream;
-import java.security.MessageDigest;
 import java.util.Map;
 import java.util.UUID;
 import java.util.function.Consumer;
 import javax.annotation.Nullable;
 
 /**
  * Simplified wrapper for computing message digests.
- *
- * @see java.security.MessageDigest
  */
 public final class Fingerprint implements Consumer<String> {
 
+  private final HashFunction hashFunction;
+  private Hasher hasher;
   // Make novel use of a CodedOutputStream, which is good at efficiently serializing data. By
   // flushing at the end of each digest we can continue to use the stream.
-  private final CodedOutputStream codedOut;
-  private final MessageDigest messageDigest;
+  private CodedOutputStream codedOut;
 
   /** Creates and initializes a new instance. */
   public Fingerprint(DigestHashFunction digestFunction) {
-    messageDigest = digestFunction.cloneOrCreateMessageDigest();
-    // This is a lot of indirection, but CodedOutputStream does a reasonable job of converting
-    // strings to bytes without creating a whole bunch of garbage, which pays off.
-    codedOut =
-        CodedOutputStream.newInstance(
-            new DigestOutputStream(ByteStreams.nullOutputStream(), messageDigest),
-            /*bufferSize=*/ 1024);
+    hashFunction = digestFunction.getHashFunction();
```

Comment:
The reason why this was reusing the MessageDigest is performance. Creating a new instance is rather expensive. If the guava functions use MessageDigest under the covers, then this could be a net loss.

Another observation: technically, Bazel doesn't *have* to use the same digest for its own action cache as for remote execution (I'm not 100% sure if Fingerprint is used anywhere in the remote execution client).